### ISSUE TITLE
out of range access

### ISSUE THREAD

**Issue by germasch** [disheng222/SZ#27]
When using SZ compression through ADIOS2, memory corruption occurs in certain cases. More specifically, trying to compress a 5x5 float array shows the issue. I don't have a simple example to reproduce the issue right now (in fact, I'm pretty sure there's more than just one problem), but if you're not aware of it, look for the address sanitizer feature of modern compilers (or use valgrind), and they will help you pinpoint and debug the problem.

The first example occurs in `sz_float.c` around line 5896. I printed `cur_data_pos - oriData` and it shows
```
cur_data_pos 6
cur_data_pos 20
cur_data_pos 18
cur_data_pos 28
=================================================================
==97516==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60b0000034f0 at pc 0x000105b66364 bp 0x7ffeebc00290 sp 0x7ffeebc00288
READ of size 4 at 0x60b0000034f0 thread T0
```
Since it's a 5x5 array, only 0..24 are valid positions. I'm pretty sure that this isn't the only problem, since the problem that started my debugging session was a write access, not a read access.

**Comment by disheng222** [comment]
Hi Kai,
Thanks for pointing out this bug!
We have fixed it in the SZ 2.1.4.2 (just released). You can also use the
master branch.

Best,
Sheng

**Comment by germasch** [comment]
Thanks, that's great! I tested it, and can confirm that the latest master in fact fixes my problem.
